Add a feature to crop the largest valid (non zero) area

yoneken opened the pull request:
This PR adds a feature to crop the largest valid (non zero) area.
The feature is useful to crop an object from the background.

For example, 

``` xml
<launch>
  <node pkg="nodelet" type="nodelet" name="standalone_nodelet" args="manager"/>

  <arg name="manager" value="standalone_nodelet" />
  <arg name="respawn" default="false" />

  <arg     if="$(arg respawn)" name="bond" value="" />
  <arg unless="$(arg respawn)" name="bond" value="--no-bond" />

  <include file="$(find openni_launch)/launch/openni.launch" />

  <!-- Crop depth images -->
  <node pkg="nodelet" type="nodelet" name="crop"
        args="load depth_image_proc/crop_foremost $(arg manager) $(arg bond)"
    respawn="$(arg respawn)" output="screen" >
    <remap from="image_raw" to="/camera/depth/image_raw" />
    <remap from="image" to="/camera/depth/image_crop" />
    <param name="distance" type="double" value="200" />
  </node>

  <!-- Crop around the center of the mass -->
  <node pkg="nodelet" type="nodelet" name="crop2"
        args="load image_proc/crop_nonZero $(arg manager) $(arg bond)"
    respawn="$(arg respawn)" output="screen" >
    <remap from="image_raw" to="/camera/depth/image_crop" />
    <remap from="image" to="/camera/depth/image_crop2" />
  </node>

  <node pkg="nodelet" type="nodelet" name="crop3"
        args="load depth_image_proc/convert_metric $(arg manager) $(arg bond)"
    respawn="$(arg respawn)" output="screen" >
    <remap from="image_raw" to="/camera/depth/image_crop2" />
    <remap from="image" to="/camera/depth/image_crop3" />
  </node>

  <node pkg="image_view" type="image_view" name="image_view">
    <remap from="image" to="/camera/depth/image" />
  </node>
  <node pkg="image_view" type="image_view" name="image_view_crop">
    <remap from="image" to="/camera/depth/image_crop3" />
  </node>

</launch>
```

![Raw image](https://pbs.twimg.com/media/CZUz7-hUkAA6POv.png)
![Cropped image](https://pbs.twimg.com/media/CZUz7TbUgAkR-J4.png)

yoneken committed:
Add a feature to crop the largest valid (non zero) area

Remove unnecessary headers

change a filename to fit for the ROS convention

vrabaud commented on image_proc/src/nodelets/crop_nonZero.cpp:
```diff
@@ -0,0 +1,147 @@
+#include <boost/thread.hpp>
+#include <cv_bridge/cv_bridge.h>
+//#include <algorithm> // for std::max_element
+#include <opencv2/opencv.hpp>
```
needed ?

vrabaud commented on image_proc/src/nodelets/crop_nonZero.cpp:
```diff
@@ -0,0 +1,147 @@
+  out_msg.encoding = raw_msg->encoding;
+  out_msg.image    = cv_ptr->image(r);
+
+  pub_.publish(out_msg.toImageMsg());
```
you can write: pub_.publish(out_msg);

yoneken replied:
When I changed to "pub_.publish(out_msg);", the following error is shown.

candidate function not viable: no known conversion from
'cv_bridge::CvImage' to 'const sensor_msgs::Image' (aka 'const
Image_std::allocator<void >') for 1st argument
void publish(const sensor_msgs::Image& message) const;

It seems that there is no feature to use CvImage directly?
https://github.com/ros-perception/image_common/blob/hydro-devel/image_transport/include/image_transport/publisher.h#L81-L89

vrabaud replied:
You're right, I thought it was a normal publisher.
https://github.com/ros-perception/vision_opencv/blob/indigo/cv_bridge/include/cv_bridge/cv_bridge.h#L298

vrabaud commented:
mostly ok but:
- should't that be in depth_image_proc instead ?
- please rename the .cpp file using the ROS conventions: http://wiki.ros.org/CppStyleGuide#Files (no capital letter)

yoneken commented:
@vrabaud 
Hi, thank you for the comment.
- I removed a line which included an unused header.
- I changed the file name to fit for the ROS convention.

> should't that be in depth_image_proc instead ?

This function is useful for general image recognizing task.
I expect that the function will be used with some OpenCV preprocessing.
This is the reason that I pushed the request to image_proc instead of depth_image_proc.

vrabaud commented:
LGTM, can you please squash the commits ? Thx

yoneken commented:
@vrabaud 
Thank you for the comment.
I squashed these commits.
